From 7e540918f5b399bfa967af11f048b7c48da4613e Mon Sep 17 00:00:00 2001
From: Han Xu <b45815@freescale.com>
Date: Wed, 21 Oct 2015 11:43:55 -0500
Subject: [PATCH 0721/5242] MLK-11747: mtd: gpmi: save bch geometry setting
 only in initial stage

commit  9fbf6c2e58906232327728982d18835c95036546 from
https://source.codeaurora.org/external/imx/linux-imx.git

fix the bch setting issue when system suspend/resume, the bch geometry
only need to be saved to debugfs in driver initial stage

Signed-off-by: Han Xu <b45815@freescale.com>
(cherry picked from commit 3b4f7178854e428fb5ef08d554b13abe4f27c533)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c  |   36 ++++++++++++++++------------
 drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c |    5 ++++
 drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.h |    1 +
 3 files changed, 27 insertions(+), 15 deletions(-)

diff --git a/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c b/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c
index 97e2285..52ada26 100644
--- a/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c
+++ b/drivers/mtd/nand/raw/gpmi-nand/gpmi-lib.c
@@ -257,6 +257,27 @@ void gpmi_dump_info(struct gpmi_nand_data *this)
 		geo->block_mark_bit_offset);
 }
 
+int bch_save_geometry(struct gpmi_nand_data *this)
+{
+	struct bch_geometry *bch_geo = &this->bch_geometry;
+	struct dentry *dbg_root;
+
+	dbg_root = debugfs_create_dir("gpmi-nand", NULL);
+	if (!dbg_root) {
+		dev_err(this->dev, "failed to create debug directory\n");
+		return -EINVAL;
+	}
+
+	dbg_bch_geo.data = (void *)bch_geo;
+	dbg_bch_geo.size = sizeof(struct bch_geometry);
+	if (!debugfs_create_blob("bch_geometry", S_IRUGO,
+				dbg_root, &dbg_bch_geo)) {
+		dev_err(this->dev, "failed to create debug bch geometry\n");
+		return -EINVAL;
+	}
+	return 0;
+}
+
 /* Configures the geometry for BCH.  */
 int bch_set_geometry(struct gpmi_nand_data *this)
 {
@@ -270,7 +291,6 @@ int bch_set_geometry(struct gpmi_nand_data *this)
 	unsigned int page_size;
 	unsigned int gf_len;
 	int ret;
-	struct dentry *dbg_root;
 
 	ret = common_nfc_set_geometry(this);
 	if (ret)
@@ -284,20 +304,6 @@ int bch_set_geometry(struct gpmi_nand_data *this)
 	page_size     = bch_geo->page_size;
 	gf_len        = bch_geo->gf_len;
 
-	dbg_root = debugfs_create_dir("gpmi-nand", NULL);
-	if (!dbg_root) {
-		dev_err(this->dev, "failed to create debug directory\n");
-		return -EINVAL;
-	}
-
-	dbg_bch_geo.data = (void *)bch_geo;
-	dbg_bch_geo.size = sizeof(struct bch_geometry);
-	if (!debugfs_create_blob("bch_geometry", S_IRUGO,
-				dbg_root, &dbg_bch_geo)) {
-		dev_err(this->dev, "failed to create debug bch geometry\n");
-		return -EINVAL;
-	}
-
 	ret = pm_runtime_get_sync(this->dev);
 	if (ret < 0) {
 		dev_err(this->dev, "Failed to enable clock\n");
diff --git a/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c b/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c
index 01dcf32..92fd590 100644
--- a/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c
+++ b/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.c
@@ -2093,6 +2093,11 @@ static int gpmi_init_last(struct gpmi_nand_data *this)
 	if (ret)
 		return ret;
 
+	/* Save the geometry to debugfs*/
+	ret = bch_save_geometry(this);
+	if (ret)
+		return ret;
+
 	/* Init the nand_ecc_ctrl{} */
 	ecc->read_page	= gpmi_ecc_read_page;
 	ecc->write_page	= gpmi_ecc_write_page;
diff --git a/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.h b/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.h
index 3f32f48..aa900b9 100644
--- a/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.h
+++ b/drivers/mtd/nand/raw/gpmi-nand/gpmi-nand.h
@@ -189,6 +189,7 @@ int start_dma_with_bch_irq(struct gpmi_nand_data *,
 int gpmi_init(struct gpmi_nand_data *);
 void gpmi_clear_bch(struct gpmi_nand_data *);
 void gpmi_dump_info(struct gpmi_nand_data *);
+extern int bch_save_geometry(struct gpmi_nand_data *);
 int bch_set_geometry(struct gpmi_nand_data *);
 int gpmi_is_ready(struct gpmi_nand_data *, unsigned chip);
 int gpmi_send_command(struct gpmi_nand_data *);
-- 
1.7.9.5

